src/Stratis.Features.FederatedPeg/CollateralChecker.cs

```diff
@@ -0,0 +1,202 @@
+            this.signals = signals;
+
+            this.cancellationSource = new CancellationTokenSource();
+            this.locker = new object();
+            this.depositsByAddress = new Dictionary<string, Money>();
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+            this.blockStoreClient = new BlockStoreClient(loggerFactory, httpClientFactory, settings.CounterChainApiPort);
+        }
+
+        public async Task InitializeAsync()
+        {
+            this.memberAddedToken = this.signals.Subscribe<FedMemberAdded>(this.OnFedMemberAdded);
+            this.memberKickedToken = this.signals.Subscribe<FedMemberKicked>(this.OnFedMemberKicked);
+
+            foreach (CollateralFederationMember federationMember in this.federationManager.GetFederationMembers().Cast<CollateralFederationMember>().Where(x => x.CollateralAmount != null && x.CollateralAmount > 0))
+                this.depositsByAddress.Add(federationMember.CollateralMainchainAddress, null);
+
+            while (true)
+            {
+                bool success = await this.UpdateCollateralInfoAsync(this.cancellationSource.Token).ConfigureAwait(false);
+
+                this.logger.LogWarning("Failed to update collateral. Ensure that mainnet gateway node is running and API is enabled. " +
+                                       "Node will not continue initialization before another gateway node responds.");
+
+                if (!success)
+                {
+                    try
+                    {
+                        await Task.Delay(CollateralInitializationUpdateIntervalSeconds * 1000, this.cancellationSource.Token).ConfigureAwait(false);
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        this.logger.LogTrace("(-)[CANCELLED]");
+                        return;
+                    }
+                }
+                else
+                {
+                    break;
+                }
+            }
+
+            this.updateCollateralContinuouslyTask = this.UpdateCollateralInfoContinuouslyAsync();
+        }
+
+        private async Task UpdateCollateralInfoContinuouslyAsync()
+        {
+            while (!this.cancellationSource.Token.IsCancellationRequested)
+            {
+                try
+                {
+                    await Task.Delay(CollateralUpdateIntervalSeconds * 1000, this.cancellationSource.Token).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException )
+                {
+                    this.logger.LogTrace("(-)[CANCELLED]");
+                    return;
+                }
+
+                bool success = await this.UpdateCollateralInfoAsync(this.cancellationSource.Token).ConfigureAwait(false);
+
+                if (!success)
+                    this.logger.LogWarning("Failed to update collateral. Ensure that mainnet gateway node is running and API is enabled.");
```

Comment:
We may need better info here (maybe not in this PR but returnig the reason of failure from `UpdateCollateralInfoAsync` can help

Reply:
The reason is being logged. 

It makes sense to return it only if we act differently based on the reason which we don't do here.

Reply:
Its being logged in trace mode so the output is not immediately visible (perhaps in that case change it to debug mode).